### ISSUE TITLE
Movie and AAC audio gain

### ISSUE THREAD

**Issue by outlyer** [mdhiggins/sickbeard_mp4_automator#219]
(I'll preface this by noting that I can propose a patch if this is desired)

Some of you may have noticed that _DTS -> AC3_ and _AC3->AAC_ often have less than ideal dynamic ranges.  This is particularly notable when playing back video on an iPad or Mac which aren't terribly loud in the first place. 

There are ways to fix this, but I'm suggested that there is a setting in the configuration file to set these options up automatically.

For example, given an _DTS->AC3_ conversion, we apply the `compand` filter:

`-filter:a:1 "compand=.3|.3:1|1:-90/-60|-60/-40|-40/-30|-20/-20:6:0:-90:0.2"`

and given a _AC3->AAC_ conversion, we boost the peak volume by a few decibels. My experience is that most of the video files I'm seeing are too quiet by between 8dB and 9dB meaning:

`-filter:a:0 "volume=8dB"`

would sufficiently normalize the volume. You can be more accurate by actually measuring the volume difference using this:

`ffmpeg -i "$1" -af "volumedetect" -f null /dev/null`

but this does take longer as it has to pass over the entire file.

So, basically, we'd have a `dts-to-ac3` setting and a `ac3-to-aac` option in the configuration file  and use this to handle the downconversions.  The actual command would be the filter commands I suggested above but could be user modified as needed.

**Comment by mdhiggins** [comment]
Lately I've been toying with the idea of trying to change how the current options are setup. The mp4 section has become cluttered with lots and lots of options and is becoming more convoluted and makes initial setup difficult. I've been toying with the idea of trying to make some kind of xml-type template that users would be able to define the standard that all their media would conform to and have the converter try to achieve that. Not 100% sure the best way to go about it or if that would actually make things easier.

Anyway, I found this script:

https://gist.github.com/slhck/99020a1a54e59cf94042

Wondering if we could adapt this and maybe just have 1 option to set a standard normalized volume?

**Comment by outlyer** [comment]
The main advantage of the current configuration file is that it is really easily edited, while XML has the potential to get tricky because of the need to open/close every tag, comply with a DTD, etc. 

I think you're 100% correct in that a revision would be really helpful, but maybe something in between the current version and XML like YAML?

As for the other issue, the trick is that gain normalization is fine for the stereo audio, but converting DTS to AC3 is a bit of a different issue; it's mainly because of how dialog is in the center channel in 5.1 audio but conversion from 5 to 2 channels leads to that dialog becoming very quiet, which is the main reason I suggested two different filters...

Before I discovered this invaluable script, I had my own more inflexible code but what it did was this

For DTS->AC3:
```sh
sox audio.wav audio-out.wav compand 0.3,1 6:-70,-60,-36 -5 -90 0.2```
```

and for AC3->AAC:
```sh
afconvert -v -d aac -b 128000 -f m4af -q 127 -s 2 audio.wav output.m4a
aacgain -r -p -t -k output.m4a output-gain.m4a
```

and then mux that all together. 

It's obviously less elegant than a single line of ffmpeg. Alternately, you could do this:

```sh

/usr/local/bin/ffmpeg -i 'input.mkv' -i 'input.eng.srt' -vcodec copy -map 0:0 -c:a:0 aac -filter:a:0 "volume=8dB" -map 0:1 -ac:a:0 2 -b:a:0 256k -metadata:s:a:0 language=eng -strict experimental -c:a:1 copy -map 0:1 -metadata:s:a:1 language=eng -c:s:0 mov_text -map 0:2 -metadata:s:s:0  language=eng -c:s:1 mov_text -map 1:0 -metadata:s:s:1 language=eng -f mp4 -threads auto -y 'output.m4v'
```

to do the normalization in the one command.

**Comment by antonino-u** [comment]
Bumping up the volume is fine, but something which helps with the dialog and whatnot is downmixing to dolby pro logic II.

Something like this... 
ffmpeg -i in.mkv
-filter_complex "[0:1]aresample=matrix_encoding=dplii[dp]"
-map 0:v -c copy
-map "[dp]" -c:a:0 libfdk_aac -vbr 5 -ac:a:0 "2"
-map 0:1 -c:a:1 copy
-movflags +faststart out.mp4

Will map the second stream in the input file to a dpl2 downmix, then write that as the first audio stream in the output alongside the copied video and copied audio.

**Comment by mdhiggins** [comment]
I'm not well versed in these issues so pardon my stupid questions. But the volume normalization issue caused by going from 5.1 to stereo where the center channel needs to be amplified, is that a specific issue when going from DTS to AAC or does that occur with any codec when going from 5.1 to stereo

**Comment by mdhiggins** [comment]
Using the ebu r128 option seems to be much faster than the "volumedetect" option

```
ffmpeg -nostats -i input.mp4 -vn -filter_complex [0:1]ebur128  -f null nul
```

About 10 seconds for a half hour file. And you can point it to specific streams which is useful. Would just need to wrap it in a python function which wouldn't be hard.

**Comment by mdhiggins** [comment]
So I've been able to incorporate into the mkvtomp4 script the ability to determine a linear gain for each audio stream based on a goal LUFS in a relatively quick manner. I just am not sure what exactly to do with that information.

It seems that the most important factor isn't the codec but instead the number of channels in the source vs the output. Seems like when converting between codecs with the same number of channels, simply using the `volume` filter is the best approach, and when converting from a higher number of channels to a lower number, then the `compand` filter is the better option. I'm not 100% clear if what I'm saying is accurate so if you could provide your input it'd be appreciated.

Here are the functions I have that determine the linear gain, adapted from the neg23 python script:
```
def r128Stats(filePath, stream=0):
    """ takes a path to an audio file, returns a dict with the loudness
    stats computed by the ffmpeg ebur128 filter """
    ffargs = [settings.ffmpeg,
              '-nostats',
              '-i',
              filePath,
              '-filter_complex',
              '[a:%s]ebur128' % stream,
              '-f',
              'null',
              '-']
    proc = subprocess.Popen(ffargs, stderr=subprocess.PIPE)
    stats = proc.communicate()[1]
    summaryIndex = stats.rfind('Summary:')
    summaryList = stats[summaryIndex:].split()
    ILufs = float(summaryList[summaryList.index('I:') + 1])
    IThresh = float(summaryList[summaryList.index('I:') + 4])
    LRA = float(summaryList[summaryList.index('LRA:') + 1])
    LRAThresh = float(summaryList[summaryList.index('LRA:') + 4])
    LRALow = float(summaryList[summaryList.index('low:') + 1])
    LRAHigh = float(summaryList[summaryList.index('high:') + 1])
    statsDict = {'I': ILufs, 'I Threshold': IThresh, 'LRA': LRA,
                 'LRA Threshold': LRAThresh, 'LRA Low': LRALow,
                 'LRA High': LRAHigh}
    return statsDict


def linearGain(iLUFS, goalLUFS=-23):
    """ takes a floating point value for iLUFS, returns the necessary
    multiplier for audio gain to get to the goalLUFS value """
    gainLog = -(iLUFS - goalLUFS)
    return 10 ** (gainLog / 20)
```

**Comment by outlyer** [comment]
Sorry it took so so long to reply.  I think you're correct that using ebu r128 is better for fixing stereo audio, particularly for music but it tends to give weird results with multi-channel.  

Downsampling is mainly a problem because DTS and AC3 tend to have wide ranges. So what we're mainly concerned about here is dynamic range compression (i.e. the difference between loudest and softest sounds) In general, the majority of TV shows could benefit from the surround stream being treated with ```compand``` and the AAC with a ```volume``` filter. 

I think adding the filter command in the config  file should work, but there needs to be a volume detection done if we want to be super accurate about getting the best (audio range) compression without clipping. 

The weird thing is that every broadcast TV program I've checked has always been too quiet by 8dB.

**Comment by mdhiggins** [comment]
How would you recommend handling the volume detection for non-stereo audio sources? The volumedetect filter?

**Comment by outlyer** [comment]
I think that is the only way to do it; AC-3 and DTS don't appear to include any sort of peak volume metadata (unlike MP3/AAC) so measuring each frame seems like the only way to accomplish it. 

That said, *some* AC-3 streams contain dialnorm and DRC flags which actually do specific the peaks. I found this on the FFMPEG lists that might illuminate how this could work:

https://ffmpeg.org/pipermail/ffmpeg-devel/2014-August/161911.html

And then you could encode with any of the following (from the same thread)
```sh
./ffmpeg -heavy_compr 1 -target_level -20 -i Broadway-5.1.ac3 -acodec
pcm_s16le -ac 2 BroadwayHeavyDRC.wav

./ffmpeg -drc_scale 1 -target_level -31 -i Broadway-5.1.ac3 -acodec
pcm_s16le -ac 2 BroadwayLightDRC.wav

./ffmpeg -drc_scale 0 -target_level -31 -i Broadway-5.1.ac3 -acodec
pcm_s16le -ac 2 BroadwayNoDRC.wav
```

**Comment by mdcollins05** [comment]
Just wanted to show my support for this feature.

I've been trying to figure out how to use ffmpeg to normalize audio (with volumedetect and ebur128) and later apply the "dynaudnorm" filter to compress the dynamic range a bit. My experience with ffmpeg is really limited so I'm stumbling through it. If I get it working, I'd make a post processing script.

I've been able to get it to normalize a single audio stream but the output is missing any other audio streams and it seems to have used a lower quality bitrate. If this feature were built in, that would be awesome.

**Comment by SgtBatten** [comment]
yep me too. starting to notice many tv shows are very quiet since using this script

**Comment by Collisionc** [comment]
I've been using the loudnorm (https://ffmpeg.org/ffmpeg-filters.html#toc-loudnorm -- effectively ebur128 ) filter for audio and while it's not perfect, it's definitely better than straight 5.1 ---> 2.0 without a filter.

**Comment by mdhiggins** [comment]
Can you post what settings you use specifically?

**Comment by Collisionc** [comment]
libfdk_aac - 2 max audio channels - 192 audio channel bitrate + default options for loudnorm. I tried to tweak the settings for loudnorm some but couldn't really make it any better.

The one thing that sucks about loudnorm is that it is a single-threaded double pass filter that takes about twice the time that dynaudnorm did. I generally copy the video stream unless it's hevc or a ludicrously high bitrate, and just passing a tv show through loudnorm/libfdk_aac takes about 2~ minutes, while movies can take up to 5 minutes.

I swapped from sabnzbd to nzbget due to this, as nzbget supports post-processing multiple files at the same time. I had to build nzbget myself and limit it to 2 post-processes because I use nvenc to re-encode videos when it is needed.

I did try dynaudnorm for a while, but the variance in volume between different episodes/tv shows was annoying. Each individual episode was fine, but due to how dynaudnorm works it led to 20-30% volume average differences between different episodes. With loudnorm on everything I almost never have to adjust the volume on my tv, and when I do it's only +/- 5%

**Comment by mdcollins05** [comment]
@Collisionc Were you able to achieve this via the built-in configuration options in the autoProcess.ini file? If so, could I get you to share your settings? I'm having problems getting it to work with the audio-filter options, specifically ffmpeg failing saying 'invalid sample rate'.

Thanks!

**Comment by mdcollins05** [comment]
It looks like I was able to get this working (well, ffmpeg doesn't complain any more and the script successfully post-processes files).

The two relevant settings in the autoProcess.ini file are:
```audio-filter = loudnorm=I=-16:TP=-1.5:LRA=11
postopts = -ar:a, 48000```

Please note that I'm not exactly sure what the values passed to loudnorm do but they were what was suggested in my search of how to use it.

**Comment by mdcollins05** [comment]
Following up with my previous comment, I'm not so sure it is working. While I'm not seeing any errors, my audio levels for videos are still all over the place.

Anyone know how I could go about confirming if this is working?

**Comment by lincolnthalles** [comment]
loudnorm works, but it is EXTREMELY slow to the point encoding to h264 is faster than it.
I'm testing pan filter right now and it is a dozen times faster than loudnorm.

`pan=stereo|c0=0.25*c0+c2+0.6*c3|c1=0.25*c1+c2+0.6*c3`

**Comment by mdcollins05** [comment]
Interesting, are you using the audio-filter and postopts configuration settings? 

I feel that mine isn't working and I'm not sure how to verify. Sound levels are quite different between shows.

**Comment by lincolnthalles** [comment]
You could try this (it is subtle, but seems to work for me)
`audio-filter = pan=stereo|c0<c0+1.414c2+.5c4+.5c6|c1<c1+1.414c2+.5c5+.5c7, volume=1.25`

or this
`audio-filter = pan=stereo|c0=c2+0.30*c0+0.30*c4|c1=c2+0.30*c1+0.30*c5`

I got them from [here](https://ffmpeg.zeranoe.com/forum/viewtopic.php?f=17&t=1851)
For some reason my ffmpeg build shows error using regular channel naming, so I changed these filters according to this mapping:
`FL=c0|FR=c1|FC=c2|LFE=c3|BL=c4|BR=c5|SL=c6|SR=c7`

In postops I only use this (it is faster than qtfaststart.py and avoids writing a new file)
`postopts = -movflags,+faststart`

**Comment by mdcollins05** [comment]
I just realized that despite my autoProcess.ini having the settings I've mentioned above, they aren't actually being put into the ffmpeg command (as seen by using `ps -ef`):

```
/home/mdcollins/ffmpeg/ffmpeg-3.4.1-64bit-static/ffmpeg -i /data/rw/media/TV/mediafile.mp4.original -vcodec copy -map 0:0 -c:a:0 copy -map 0:1 -metadata:s:a:0 language=und -disposition:a:0 default -c:a:1 copy -map 0:2 -metadata:s:a:1 language=und -disposition:a:1 none -f mp4 -threads auto -ar:a 48000 -maxrate 10000k -bufsize 30000k -y /data/rw/media/TV/mediafile.mp4
```

I am, however, reprocessing old MP4 files, so I don't know if that'll make a difference.

**Comment by mdcollins05** [comment]
I've discovered that my audio-filter does work for files that are not already an mp4.

```ffmpeg -i /data/rw/media/TV/another_show.mkv -vcodec libx264 -map 0:0 -crf 21 -c:a:0 ac3 -map 0:1 -disposition:a:0 default -ac:a:0 2 -b:a:0 512k -filter:a:0 loudnorm=i=-16:tp=-1.5:lra=11 -metadata:s:a:0 language=eng -f mp4 -threads auto -ar:a 48000 -maxrate 10000k -bufsize 30000k -y /data/rw/media/TV/another_show.mp4```

This one, for whatever reason, wasn't converted previously.

Is there a way to force the audio-filter to be applied to all files? I'm currently manually running the process via a bash loop on older files (to extract all subtitles), like so:

`./manual.py -cmp4 -i "$file"`

**Comment by lincolnthalles** [comment]
Your problem is not in reprocessing mp4 files, it is that audio is already in desired format. Filters are not applied by ffmpeg on any stream **copy** operation.

Your manual.py output shows
`"codec": "copy",`
And your ffmpeg command line
`-c:a:0 copy`
`-c:a:1 copy`

Assuming you are trying to recreate AAC Stereo tracks to apply audio filtering AND you have kept original audio tracks (DTS, EAC3 or AC3) on your old mp4 files, you must process them **2 times**:

- Edit autoProcess.ini:
```
ios-audio = False
audio-codec =dts,eac3,ac3
max-audio-channels =
audio-copy-original = False
```
- Run manual.py:
This settings will discard existing AAC tracks and will generate new mp4's containing only original multichannel tracks.

- Switch back autoProcess.ini to your previous audio settings: 
```
ios-audio = False
audio-codec = libfdk_aac,aac
max-audio-channels = 2
audio-copy-original = True
audio-channel-bitrate = 107
audio-filter = pan=stereo|c0<c0+1.414c2+.5c4+.5c6|c1<c1+1.414c2+.5c5+.5c7, volume=1.25
```
- Run manual.py on recently generated mp4's:
At this time, audio filtering will be applied (as there is no existing aac stereo tracks).

**Comment by mdcollins05** [comment]
I thought I'd give an update on this. I ended up using ffmpeg-normalize (https://github.com/slhck/ffmpeg-normalize) and wrote a post process script (https://github.com/mdcollins05/ffmpeg-normalize-extras) for sickbeard_mp4_automator. This allowed me to easily process my existing library. My only "issue" with it (and it's not a big deal at all) is that my post process script seems to happen after the file is moved to it's final location and not processed while still in nzbget.

**Comment by canadabri** [comment]
I'd like to add this to my autoprocess.ini   Is there info on how to do this - i.e. install instructions ?  My TV setup at home is a TV sound bar and a subwoofer (2.1 ?) .  This thread explains why the audio seems so low on some videos I have been watching

**Comment by Maximus97** [comment]
So `pan=stereo|c0<c0+1.414c2+.5c4+.5c6|c1<c1+1.414c2+.5c5+.5c7, volume=1.25` is what you concluded was the best? Or do you use another audio filter?

**Comment by stevemk14ebr** [comment]
I think this could be used: https://github.com/lordmulder/DynamicAudioNormalizer . It mentions it's available as an ffmpeg filter and supports lots of formats.

**Comment by mdhiggins** [comment]
Yeah I think `DYNAUDNORM` is the answer here
